Импортируем библиотеки

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import networkx as nx


Считываем файл

In [205]:
data = pd.read_excel('test.xlsx')
data.head(10)


Provider ID    Operation ID  Payment ID Operation Type Operation Status  \
0       4781.0  99892010097211  6021093719           sale          success   
1       1293.0  99888010096111   347788572           sale          success   
2       1293.0  99888010096101   347788572           sale          decline   
3       1293.0  99888010096091   347788572           sale          decline   
4       1293.0  99888010096081   347788572           sale          decline   
5       1293.0  99888010096071   347788572           sale          decline   
6       1293.0  99888010096061   347788572           sale          decline   
7       1293.0  99888010096051   347788572           sale          decline   
8       1293.0  99888010096041   347788572           sale          decline   
9       1293.0  99888010096031   347788572           sale          decline   

   Secure 3ds Check  Channel Amount Channel Currency  \
0               1.0          1000.0              RUB   
1               NaN            51.0              EUR   
2               NaN            51.0              EUR   
3               NaN            51.0              EUR   
4               NaN            51.0              EUR   
5               NaN            51.0              EUR   
6               NaN            51.0              EUR   
7               NaN            51.0              EUR   
8               NaN            51.0              EUR   
9               NaN            51.0              EUR   

  Operation created at Date Payment Method Type    Account Number  \
0                2022-03-30                visa  427626******5062   
1                2022-02-27                visa  434597******0730   
2                2022-02-27                visa  434597******0730   
3                2022-02-27                visa  434597******0730   
4                2022-02-27                visa  434597******0730   
5                2022-02-27                visa  434597******0730   
6                2022-02-27                visa  434597******0730   
7                2022-02-27                visa  434597******0730   
8                2022-02-27                visa  434597******0730   
9                2022-02-27                visa  434597******0730   

   Merchant ID  Project ID Customer ID                   Customer Email  \
0           39         321     3753071                 avtomr@yandex.ru   
1         1581        5771       95022  williahmalexdelaplaza@gmail.com   
2         1581        5771       95022  williahmalexdelaplaza@gmail.com   
3         1581        5771       95022  williahmalexdelaplaza@gmail.com   
4         1581        5771       95022  williahmalexdelaplaza@gmail.com   
5         1581        5771       95022  williahmalexdelaplaza@gmail.com   
6         1581        5771       95022  williahmalexdelaplaza@gmail.com   
7         1581        5771       95022  williahmalexdelaplaza@gmail.com   
8         1581        5771       95022  williahmalexdelaplaza@gmail.com   
9         1581        5771       95022  williahmalexdelaplaza@gmail.com   

                 Card Holder Issuer Country Name  Country By IP Name  \
0           LEONID PROKHOROV  Russian Federation  Russian Federation   
1  WILLIAHM ALEX DE LA PLAZA             Denmark             Denmark   
2  WILLIAHM ALEX DE LA PLAZA             Denmark             Denmark   
3  WILLIAHM ALEX DE LA PLAZA             Denmark             Denmark   
4  WILLIAHM ALEX DE LA PLAZA             Denmark             Denmark   
5  WILLIAHM ALEX DE LA PLAZA             Denmark             Denmark   
6  WILLIAHM ALEX DE LA PLAZA             Denmark             Denmark   
7  WILLIAHM ALEX DE LA PLAZA             Denmark             Denmark   
8  WILLIAHM ALEX DE LA PLAZA             Denmark             Denmark   
9  WILLIAHM ALEX DE LA PLAZA             Denmark             Denmark   

  Customer IP V4 V6    Issuer Bank Name  
0    176.59.134.184  SBERBANK of Russia  
1    62.198.132.139      Jyske Bank A/S  
2    62.198.132.139   

In [202]:
data.isnull().sum()



Provider ID                    355
Operation ID                     0
Payment ID                       0
Operation Type                   0
Operation Status                 0
Secure 3ds Check             12366
Channel Amount                   3
Channel Currency                13
Operation created at Date        0
Payment Method Type              7
Account Number                 375
Merchant ID                      0
Project ID                       0
Customer ID                     33
Customer Email                1495
Card Holder                   1151
Issuer Country Name           1062
Country By IP Name              56
Customer IP V4 V6                0
Issuer Bank Name              1176
dtype: int64

In [214]:
data_has_null = data.drop(columns=['Secure 3ds Check', 'Customer Email'])
data_has_null = data_has_null[data_has_null.isnull().any(axis=1)]
agg_na = data_has_null.groupby(['Merchant ID', 'Project ID']).agg(
    total=('Customer ID', 'size'),
    total_decline=('Operation Status', lambda x: (x == 'decline').sum()),
    unique_customers=('Customer ID', 'nunique'),
    unique_IPs=('Customer IP V4 V6', 'nunique'),
    unique_Operations=('Operation ID', 'nunique'),
    
    unique_acc_number=('Account Number', 'nunique'),
    unique_cardholders=('Card Holder', 'nunique'),
    total_country_missmatch=('Country By IP Name', lambda x: (x != data.loc[x.index, 'Issuer Country Name']).sum()),
).reset_index()
agg_na = agg_na.sort_values(by='total', ascending=False)
# .div(agg_projects['total'], axis=0).mul(100).round()
# try:
#     agg_projects[columns_to_convert] = agg_projects[columns_to_convert].applymap(lambda x: f'{int(x)}%')
# except:
#     pass

agg_na

Merchant ID  Project ID  total  total_decline  unique_customers  \
7           278         983    475            146               443   
2            46         433    326            252               137   
0            39         321    326            115               106   
3           212         726    175            175               170   
1            46         162    123             59                20   
11         1581        5771     89             88                 4   
10          300        1098     71             38                56   
8           278       22431     17             17                 1   
4           212        5591      9              9                 1   
6           212       30791      8              6                 4   
5           212       12571      7              7                 6   
9           278       26751      5              5                 3   
12         2901       18981      1              1                 1   
13         4011       40571      1              1                 1   
14         6841       25211      1              1                 1   

    unique_IPs  unique_Operations  unique_acc_number  unique_cardholders  \
7          440                475                410                   3   
2          161                326                 72                  45   
0          121                326                102                  12   
3            2                175                169                   1   
1           22                123                  0                   0   
11           6                 89                 80                  87   
10          58                 71                 57                   0   
8            4                 17                 15                   1   
4            2                  9                  1                   1   
6            4                  8                  4                   4   
5            5                  7                  5                   5   
9            3                  5                  3                   3   
12           1                  1                  1                   1   
13           1                  1                  1                   1   
14           1                  1                  1                   1   

    total_country_missmatch  
7                       414  
2                       304  
0                       229  
3                       175  
1                       123  
11                       88  
10                       42  
8                        16  
4                         0  
6                         8  
5                         0  
9                         0  
12                        1  
13                        0  
14                        0

In [233]:
agg_data = data.groupby(['Card Holder']).agg(
    total=('Customer ID', 'size'),
    total_decline=('Operation Status', lambda x: (x == 'decline').sum()),
    unique_acc_number=('Account Number', 'nunique'),
    unique_ips=('Customer IP V4 V6', 'nunique'),
    unique_cardholders=('Card Holder', 'nunique'),
    unique_emails=('Customer Email', 'nunique')
).reset_index()
agg_data = agg_data.sort_values(by='unique_acc_number', ascending=False)

fraud_lst = ['ALEKSEY SHISHKIN', 'JOHN DOE' , 'NO NAME']
data_fraud1 = data[data['Card Holder'].isin(fraud_lst)]
data_fraud2 = data[data['Customer IP V4 V6'] == '127.0.0.1']
agg_data.head(10)

Card Holder  total  total_decline  unique_acc_number  unique_ips  \
269     ALEKSEY SHISHKIN    202            202                184           4   
1909            JOHN DOE    277            111                155         170   
2594             NO NAME     36             36                 17           5   
1989        KAPITAL BANK     28             20                 11          13   
2422          MOMENTUM R     13              6                  6           6   
3862                VISA     13             10                  5           5   
2382     MICHAEL ONWURAH     23             20                  4          16   
2689  OLEGARIO RODRIGUEZ     10             10                  4           3   
666        AZEEZ ODUTOLA      4              3                  3           2   
695     BAHRUZ ISMAYILOV      7              7                  3           1   

      unique_cardholders  unique_emails  
269                    1              1  
1909                   1              2  
2594                   1              1  
1989                   1              1  
2422                   1              6  
3862                   1              1  
2382                   1              1  
2689                   1              2  
666                    1              1  
695                    1              1

In [235]:
agg_data = data.groupby(['Customer IP V4 V6']).agg(
    total=('Customer ID', 'size'),
    total_decline=('Operation Status', lambda x: (x == 'decline').sum()),
    unique_acc_number=('Account Number', 'nunique'),
    unique_countries=('Country By IP Name', 'nunique'),
    unique_cardholders=('Card Holder', 'nunique'),
    unique_emails=('Customer Email', 'nunique')
).reset_index()
agg_data = agg_data.sort_values(by='unique_acc_number', ascending=False)
fraud_lst2 = agg_data.head(20)['Customer IP V4 V6']
data_fraud3 = data[data['Card Holder'].isin(fraud_lst2)]
agg_data.head(20)

Customer IP V4 V6  total  total_decline  unique_acc_number  \
3260        5.62.19.48    242            242                192   
3258        5.62.18.88    123            123                112   
3257        5.62.18.60     75             75                 69   
4767     95.174.65.147    174            171                 55   
3617       80.255.7.20     33             33                 33   
3264       5.62.62.185     42             42                 28   
3263       5.62.60.194     49             49                 26   
3614     80.255.13.117     29             29                 21   
3262       5.62.58.121     31             31                 15   
3768      83.142.83.20     29             29                 13   
4766     95.174.65.131     22             22                 12   
2567       45.45.45.45     17              6                  9   
3275        5.8.16.147     69             64                  7   
3276        5.8.16.148     60             57                  6   
20      102.176.65.201      7              7                  6   
3947    85.132.106.196      5              2                  5   
3115       5.44.37.187     17             14                  5   
3618        80.255.7.4      8              8                  4   
1394   190.236.156.155      9              8                  4   
407       154.160.5.70      4              4                  4   

      unique_countries  unique_cardholders  unique_emails  
3260                 1                 195              1  
3258                 1                   1              1  
3257                 1                   1              1  
4767                 1                  60              1  
3617                 1                  33              1  
3264                 1                  28              1  
3263                 1                  24              1  
3614                 1                  21              1  
3262                 1                  15              1  
3768                 1                  13              1  
4766                 1                   1              1  
2567                 1                   3              0  
3275                 1                   7              1  
3276                 1                   6              1  
20                   1                   0              1  
3947                 1                   5              2  
3115                 1                   5              1  
3618                 1                   4              1  
1394                 1                   3              1  
407                  1                   0              1

In [105]:
data_bank_missing = data[data['Issuer Country Name'].isna()]
missed_merch = data_bank_missing['Merchant ID'].unique().tolist()
agg_data = data_bank_missing.groupby(['Merchant ID', 'Project ID']).agg(
    total=('Customer ID', 'size'),
    total_decline=('Operation Status', lambda x: (x == 'decline').sum()),
    unique_acc_number=('Account Number', 'nunique'),
    unique_countries=('Country By IP Name', 'nunique'),
    unique_cardholders=('Card Holder', 'nunique'),
    unique_emails=('Customer Email', 'nunique')
).reset_index()
agg_data = agg_data.sort_values(by='total', ascending=False)
agg_data.head()



Merchant ID  Project ID  total  total_decline  unique_acc_number  \
3          278         983    414            112                361   
2           46         433    263            193                 26   
0           39         321    218             23                 90   
1           46         162    123             59                  0   
4          300        1098     42             10                 41   

   unique_countries  unique_cardholders  unique_emails  
3                 3                   0            351  
2                40                   1            108  
0                13                   0             18  
1                 4                   0             20  
4                 2                   0              0

In [189]:
data_bank_missing = data[data['Issuer Country Name'].isna()]
data_bank_missing.head(10)

Provider ID    Operation ID  Payment ID Operation Type Operation Status  \
18        1083.0  99817010094891      299486           sale          decline   
23        1178.0   9967010044951      292536           sale          decline   
46        2791.0   9938010044521      294668           sale          decline   
130       2801.0  98220010095101      294667           sale          decline   
163       3211.0  97217010094201      292068           sale          decline   
165       1521.0  97124010096841  6021018333           sale          success   
191       1521.0  96443010098511  6021159272           sale          success   
203       1083.0  95968010095941      298344           sale          decline   
216       1083.0  95520010094401      294728           sale          decline   
230       3211.0  95051010094091      294221           sale          decline   

     Secure 3ds Check  Channel Amount Channel Currency  \
18                NaN           17.70              USD   
23                NaN          200.00              MYR   
46                NaN          600.00              THB   
130               NaN          600.00              THB   
163               NaN          100.00              MYR   
165               NaN        20000.00              UZS   
191               NaN       100000.00              UZS   
203               NaN          168.17              USD   
216               NaN           10.00              USD   
230               NaN          100.00              MYR   

    Operation created at Date Payment Method Type  \
18                 2022-02-27      webmoney-light   
23                 2022-02-11      banks/malaysia   
46                 2022-02-17      banks/thailand   
130                2022-02-17       banks/thai-qr   
163                2022-02-10      banks/malaysia   
165                2022-03-30        card-partner   
191                2022-03-30        card-partner   
203                2022-02-25      webmoney-light   
216                2022-02-17      webmoney-light   
230                2022-02-15      banks/malaysia   

                            Account Number  Merchant ID  Project ID  \
18                                     NaN           46         433   
23                                     NaN           46         162   
46                                     NaN           46         433   
130                                    NaN           46         433   
163                                    NaN           46         162   
165  1521-703cd2530dc5d1ffa84a006bf39b5a1e           39         321   
191  1521-487fa1193844a877d70757192afa6aa9           39         321   
203                                    NaN           46         433   
216                                    NaN           46         433   
230                                    NaN           46         162   

    Customer ID              Customer Email Card Holder Issuer Country Name  \
18       355878    chemutairuth20@gmail.com         NaN                 NaN   
23       284261          joelai90@yahoo.com         NaN                 NaN   
46       352624   kolin0949281008@gmail.com         NaN                 NaN   
130      352624   kolin0949281008@gmail.com         NaN                 NaN   
163      284261          joelai90@yahoo.com         NaN                 NaN   
165     4639134                         NaN         NaN                 NaN   
191     4054450                         NaN         NaN                 NaN   
203      289753  torben_schaefer@hotmail.de         NaN                 NaN   
216      352693     dellingsalpha@gmail.com         NaN                 NaN   
230      352003       zulinah1998@gmail.com         NaN                 NaN   

    Country By IP Name Customer IP V4 V6 Issuer Bank Name  
18               Kenya    105.48.159.220              NaN  
23            Malaysia   202.186.143.202              NaN  
46            Thailand       1.47.27.183              NaN  
130      

In [104]:
data[data['Merchant ID'].isin(missed_merch)]['Project ID'].unique().tolist()

[321, 5771, 433, 162, 22431, 26751, 983, 1098, 5741, 5761]

In [167]:

agg = data.groupby(['Payment Method Type']).size().reset_index(name='counts')

counts_sorted = agg.sort_values(by='counts', ascending=False, inplace=True)
agg.head(10)

Payment Method Type  counts
34                visa    8811
28          mastercard    3396
29                 mir     269
2              BEELINE     198
5                TELE2     135
14      banks/malaysia     133
39      webmoney-light     128
20        card-partner      94
31             monetix      81
9    bank-transfer/upi      64

In [154]:
agg_merchants = data.groupby('Merchant ID').agg(
    total=('Customer ID', 'size'),
    total_decline=('Operation Status', lambda x: (x == 'decline').sum()),
    unique_IPs=('Customer IP V4 V6', 'nunique'),
    unique_acc_number=('Account Number', 'nunique'),
    unique_cardholders=('Card Holder', 'nunique'),
    unique_emails=('Customer Email', 'nunique'),
    total_country_missmatch=('Country By IP Name', lambda x: (x != data.loc[x.index, 'Issuer Country Name']).sum())
).reset_index()
agg_merchants = agg_merchants.sort_values(by='total', ascending=False)

columns_to_convert = agg_merchants.columns.difference(['Merchant ID', 'total'])

agg_merchants[columns_to_convert] = agg_merchants[columns_to_convert]
# .div(agg_merchants['total'], axis=0).mul(100).round()
# agg_merchants[columns_to_convert] = agg_merchants[columns_to_convert].applymap(lambda x: f'{int(x)}%')

agg_merchants.head(10)



Merchant ID  total  total_decline  unique_IPs  unique_acc_number  \
5         1581   2947           2690         231                468   
1           46   2795           1874         923                656   
2          212   2675           1714        1118               1258   
0           39   2500           1202         952                853   
3          278   2438           1378        1757               1715   
4          300    159            121          75                103   
8         6841     36             36           3                 36   
6         2901     22             22           1                 12   
7         4011      3              3           1                  2   

   unique_cardholders  unique_emails  total_country_missmatch  
5                 473            173                      762  
1                 599            618                      900  
2                 958            109                      238  
0                 744            753                      456  
3                1331           1048                      445  
4                  45              5                      113  
8                  34              1                       35  
6                   1              1                       22  
7                   1              1                        2

In [217]:
agg_projects = data.groupby(['Merchant ID', 'Project ID']).agg(
    total=('Customer ID', 'size'),
    total_decline=('Operation Status', lambda x: (x == 'decline').sum()),
    unique_customers=('Customer ID', 'nunique'),
    unique_IPs=('Customer IP V4 V6', 'nunique'),
    unique_Operations=('Operation ID', 'nunique'),
    
    unique_acc_number=('Account Number', 'nunique'),
    unique_cardholders=('Card Holder', 'nunique'),
    unique_emails=('Customer Email', 'nunique'),
    total_country_missmatch=('Country By IP Name', lambda x: (x != data.loc[x.index, 'Issuer Country Name']).sum()),
    total_3ds=('Secure 3ds Check', lambda x: (x == 1.0).sum()),
).reset_index()
agg_projects = agg_projects.sort_values(by='total', ascending=False)

columns_to_convert = agg_merchants.columns.difference(['Merchant ID', 'total'])

agg_projects[columns_to_convert] = agg_projects[columns_to_convert]
# .div(agg_projects['total'], axis=0).mul(100).round()
# try:
#     agg_projects[columns_to_convert] = agg_projects[columns_to_convert].applymap(lambda x: f'{int(x)}%')
# except:
#     pass

agg_projects.head(20)


Merchant ID  Project ID  total  total_decline  unique_customers  \
25         1581        5771   2877           2642               162   
2            46         433   2672           1815               616   
0            39         321   2500           1202               828   
11          212       12571   2013           1311               837   
19          278         983   1547           1022              1100   
20          278       22431    688            202               580   
13          212       27361    231             79               146   
21          278       26751    203            154               121   
3           212         726    177            177               172   
22          300        1098    159            121                70   
1            46         162    123             59                20   
8           212       10671     98             49                60   
23         1581        5741     61             42                11   
18          212       46311     48             12                31   
28         6841       25211     36             36                 1   
12          212       27331     28             19                11   
5           212        5591     27             23                10   
26         2901       18981     22             22                 1   
6           212        7551     18             18                15   
15          212       30791     10              7                 6   

    unique_IPs  unique_Operations  unique_acc_number  unique_cardholders  \
25         216               2877                455                 461   
2          909               2672                656                 599   
0          952               2500                853                 744   
11         833               2013                814                 831   
19        1104               1547               1071                 671   
20         583                688                583                 580   
13         153                231                138                   1   
21         120                203                128                 120   
3            4                177                171                   3   
22          75                159                103                  45   
1           22                123                  0                   0   
8           57                 98                 56                  60   
23          14                 61                 13                  12   
18          29                 48                 29                  27   
28           3                 36                 36                  34   
12          10                 28                 10                   3   
5           11                 27                 10                   8   
26           1                 22                 12                   1   
6           14                 18                 15                  12   
15           6                 10                  6                   6   

    unique_emails  total_country_missmatch  total_3ds  
25            162                      747          0  
2             606                      777          0  
0             753                      456        894  
11              1                       29          0  
19            351                      414          0  
20            580                       21        220  
13              1                        8          0  
21            121                       10         41  
3               3                      175          0  
22              5                      113          0  
1              20                      123          0  
8              61                        4          0  
23             11                       11          0  
18             27                        2          0  
28              1                       35          0  
12              1  

In [177]:
def filter_users_more_than_once(df, user_column):
    user_counts = df[user_column].value_counts()
    users_to_keep = user_counts[user_counts == 1].index
    filtered_df = df[df[user_column].isin(users_to_keep)]

    return filtered_df

data_5771 = data[data['Project ID'] == 5771]
data_5771_filtered = filter_users_more_than_once(data_5771, 'Customer ID')
len(data_5771_filtered)


28

# Что стоит отметить из аномального:
* у проекта 726 все операции отклонились, при этом на 171 номер аккаунта всего 4 IP адреса. Выглядит как мошенничнские действия.
* аналогично у проекта 18981 все операции были отклонены, при этом номеров карты 12, а уникальных ip, имен кардхолдера и емаилв всего один
* у самого большого проекта 5771 очнь высокая доля операции типа decline по сравнению с остальными, мало уникальных ip адресов, кардхолдеров и емайлов

In [134]:
agg = data.groupby('Customer IP V4 V6')['Customer ID'].nunique().reset_index(name='counts')

counts_sorted = agg.sort_values(by='counts', ascending=False, inplace=True)
agg.head(10)


Customer IP V4 V6  counts
3258        5.62.18.88     104
3257        5.62.18.60      69
2567       45.45.45.45       8
5032     95.86.178.234       6
3115       5.44.37.187       5
3947    85.132.106.196       5
1178      185.32.44.98       4
3006     5.197.226.121       4
3203       5.44.39.167       4
1176     185.32.44.122       3

In [215]:
agg = data.groupby('Operation Type').size().reset_index(name='counts')

counts_sorted = agg.sort_values(by='counts', ascending=False, inplace=True)
agg.head(10)

Operation Type  counts
6             sale   11317
3           payout    2199
4        recurring      26
0       commission      19
1  customer action      10
2    manual change       3
5           refund       1

In [114]:
data[(data['Operation Type'] != 'payout') & (data['Operation Type'] != 'sale')]

Provider ID      Operation ID  \
240         4971.0    94739010097691   
258         1095.0    93885010093551   
296         1095.0    92558010096161   
376         1095.0    90789010096781   
549         4971.0    87364010097591   
810         1421.0    82975010097781   
971         4971.0    80605010097001   
1010        4971.0    80316010097281   
1144        1095.0    78449010096471   
1170        4971.0    78118010096751   
1180        1421.0    78024010096831   
1182        1421.0    77996010097641   
1337        4971.0    76082010098131   
1347        1082.0    75853010094211   
1400        1293.0    74933010095261   
1514        4971.0    73372010097721   
1530        1095.0    73291010095231   
1620        4971.0    72488010096931   
1629        5931.0    72357010097521   
1633        4971.0    72315010097841   
1779        1095.0    69941010088871   
1944        1421.0    68407010090641   
1962        5931.0    68134010092041   
2004        4971.0    67691010090561   
2324        4971.0    63836010091351   
2507        1095.0    61541010089841   
2751        1095.0    58358010086701   
2764        1095.0    58174010088641   
3057        1095.0    54418010086871   
3275        1421.0    51889010090431   
3285        4971.0    51750010091111   
4494           NaN  5044436010102462   
5541        4971.0  5028733010098562   
5727        4971.0  5026649010098942   
7339           NaN  5004597010100112   
7399        4971.0  5003889010098842   
7581        4971.0  5001490010098532   
7981        1095.0    47250010059851   
8032        4971.0    46720010062131   
8568        1082.0    40710010058311   
8592        1421.0    40380010062321   
8754        4971.0    38324010048041   
9128        1421.0    33676010048241   
9392        4971.0    29882010048431   
9407        4971.0    29493010048161   
9567        1082.0    28112010045341   
9589        1095.0    27878010046711   
9886        1095.0    25125010044341   
9927        5931.0    24655010047961   
9955        1095.0    24201010046931   
10157       1421.0    21806010048151   
10158       4971.0    21782010047361   
10214       1095.0    21168010046831   
10336       1421.0    19574010047511   
10341          NaN    19453010048751   
10523       1095.0    17394010045541   
10774       1421.0    14268010047211   
10927       4971.0    12031010047261   
11005       4971.0    10725010047661   

                                           Payment ID   Operation Type  \
240                  2521af6bd6fe2da127d24b75017c9c82        recurring   
258    capitaller-transfer-b0462e3ce0a6e03d38a8614ade       commission   
296    capitaller-transfer-111e2bc075f4fc5035d80ddc53       commission   
376    capitaller-transfer-8e0d62ada09bbe339329f50a03       commission   
549                  4ae4ad6a0322868aac73af8f57a65393        recurring   
810                                        D318990709  customer action   
971                  f0e2c71d82cdc1c7c2370e74c139b852        recurring   
1010                 008cd4eb0082c9480d850d06138db736        recurring   
1144   capitaller-transfer-77b17ef6ada45d43cece7705ac       commission   
1170                 667c73b7ebcfe67de0f21ca4f5fabe67        recurring   
1180                                       D319016023  customer action   
1182                                       D319023317  customer action   
1337                 3d393ac7676a36bd7385af6a20c32cb8        recurring   
1347                                           290840       commission   
1400                                           273506           refund   
1514                 3bcc412f52f9d0cbcdd19b220c6eee84        recurring   
1530   capitaller-transfer-5bb0abf577b1e626e2beb4f6fc       commission   
1620                 5dd42ee364e57ee85c52d14534d53953        recurring   
1629                 06a05ad82ce3b2f9f5388bafef751ac4        recurring   
1633                 97dece09b925d49bb560fcbf3ba6d29e        recurring   
1779   capitaller-transfer-39212a104d88a1b9bf606dd54

In [116]:
data[data['Customer IP V4 V6'] == '127.0.0.1']

Provider ID      Operation ID  \
258         1095.0    93885010093551   
259         1095.0    93885010093541   
296         1095.0    92558010096161   
297         1095.0    92558010096151   
376         1095.0    90789010096781   
377         1095.0    90789010096771   
1144        1095.0    78449010096471   
1145        1095.0    78449010096461   
1530        1095.0    73291010095231   
1531        1095.0    73291010095221   
1779        1095.0    69941010088871   
1780        1095.0    69941010088861   
2507        1095.0    61541010089841   
2508        1095.0    61541010089831   
2751        1095.0    58358010086701   
2752        1095.0    58358010086691   
2764        1095.0    58174010088641   
2765        1095.0    58174010088631   
3057        1095.0    54418010086871   
3058        1095.0    54418010086861   
4034        4781.0  5051258010101182   
7981        1095.0    47250010059851   
7982        1095.0    47250010059841   
9589        1095.0    27878010046711   
9590        1095.0    27878010046701   
9886        1095.0    25125010044341   
9887        1095.0    25125010044331   
9955        1095.0    24201010046931   
9956        1095.0    24201010046921   
10214       1095.0    21168010046831   
10215       1095.0    21168010046821   
10523       1095.0    17394010045541   
10524       1095.0    17394010045531   

                                           Payment ID Operation Type  \
258    capitaller-transfer-b0462e3ce0a6e03d38a8614ade     commission   
259    capitaller-transfer-b0462e3ce0a6e03d38a8614ade         payout   
296    capitaller-transfer-111e2bc075f4fc5035d80ddc53     commission   
297    capitaller-transfer-111e2bc075f4fc5035d80ddc53         payout   
376    capitaller-transfer-8e0d62ada09bbe339329f50a03     commission   
377    capitaller-transfer-8e0d62ada09bbe339329f50a03         payout   
1144   capitaller-transfer-77b17ef6ada45d43cece7705ac     commission   
1145   capitaller-transfer-77b17ef6ada45d43cece7705ac         payout   
1530   capitaller-transfer-5bb0abf577b1e626e2beb4f6fc     commission   
1531   capitaller-transfer-5bb0abf577b1e626e2beb4f6fc         payout   
1779   capitaller-transfer-39212a104d88a1b9bf606dd547     commission   
1780   capitaller-transfer-39212a104d88a1b9bf606dd547         payout   
2507   capitaller-transfer-bbb47c85a3613ba439400b4721     commission   
2508   capitaller-transfer-bbb47c85a3613ba439400b4721         payout   
2751   capitaller-transfer-101954a1d897132f64ab776e14     commission   
2752   capitaller-transfer-101954a1d897132f64ab776e14         payout   
2764   capitaller-transfer-4055b1d42bf745475529757c9e     commission   
2765   capitaller-transfer-4055b1d42bf745475529757c9e         payout   
3057   capitaller-transfer-835da9b155ab24b27e7759573e     commission   
3058   capitaller-transfer-835da9b155ab24b27e7759573e         payout   
4034                                       6021109000         payout   
7981   capitaller-transfer-bcbfe70668898425934e2a18f6     commission   
7982   capitaller-transfer-bcbfe70668898425934e2a18f6         payout   
9589   capitaller-transfer-0cd87dd710c12b5fd05ae8eba5     commission   
9590   capitaller-transfer-0cd87dd710c12b5fd05ae8eba5         payout   
9886   capitaller-transfer-36745fcf73e32e9da2c891d8a2     commission   
9887   capitaller-transfer-36745fcf73e32e9da2c891d8a2         payout   
9955   capitaller-transfer-a12b493e9e86f7eba057a5495f     commission   
9956   capitaller-transfer-a12b493e9e86f7eba057a5495f         payout   
10214  capitaller-transfer-10d2c67626d8074578ee4fc7a5     commission   
10215  capitaller-transfer-10d2c67626d8074578ee4fc7a5         payout   
10523  capitaller-transfer-e3869ab86c9d1e90961a44d8e6     commission   
10524  capitaller-transfer-e3869ab86c9d1e90961a44d8e6         payout   

      Operation Status  Secure 3ds Check  Channel Amount Channel Currency  \
258            success               NaN            1.86              USD   
259            success               NaN   

In [78]:
import requests
def get_exchange_rates():
    with open('token.txt', 'r') as file:
        api_key = file.read().strip()
    url = f'http://data.fixer.io/api/latest?access_key={api_key}'
    response = requests.get(url)
    data = response.json()
    if 'rates' in data:
        rates = data['rates']
        rates[data['base']] = 1.0
        for d in rates:
            rates[d] = 1.0/rates[d]
        return rates
    
rates = get_exchange_rates()

data['channel_usd_value'] = data['Channel Currency'].map(rates) * data['Channel Amount']


Provider ID    Operation ID  Payment ID Operation Type Operation Status  \
0       4781.0  99892010097211  6021093719           sale          success   
1       1293.0  99888010096111   347788572           sale          success   
2       1293.0  99888010096101   347788572           sale          decline   
3       1293.0  99888010096091   347788572           sale          decline   
4       1293.0  99888010096081   347788572           sale          decline   

   Secure 3ds Check  Channel Amount Channel Currency  \
0               1.0          1000.0              RUB   
1               NaN            51.0              EUR   
2               NaN            51.0              EUR   
3               NaN            51.0              EUR   
4               NaN            51.0              EUR   

  Operation created at Date Payment Method Type  ... Merchant ID  Project ID  \
0                2022-03-30                visa  ...          39         321   
1                2022-02-27                visa  ...        1581        5771   
2                2022-02-27                visa  ...        1581        5771   
3                2022-02-27                visa  ...        1581        5771   
4                2022-02-27                visa  ...        1581        5771   

   Customer ID                   Customer Email                Card Holder  \
0      3753071                 avtomr@yandex.ru           LEONID PROKHOROV   
1        95022  williahmalexdelaplaza@gmail.com  WILLIAHM ALEX DE LA PLAZA   
2        95022  williahmalexdelaplaza@gmail.com  WILLIAHM ALEX DE LA PLAZA   
3        95022  williahmalexdelaplaza@gmail.com  WILLIAHM ALEX DE LA PLAZA   
4        95022  williahmalexdelaplaza@gmail.com  WILLIAHM ALEX DE LA PLAZA   

  Issuer Country Name  Country By IP Name Customer IP V4 V6  \
0  Russian Federation  Russian Federation    176.59.134.184   
1             Denmark             Denmark    62.198.132.139   
2             Denmark             Denmark    62.198.132.139   
3             Denmark             Denmark    62.198.132.139   
4             Denmark             Denmark    62.198.132.139   

     Issuer Bank Name channel_usd_value  
0  SBERBANK of Russia          9.554105  
1      Jyske Bank A/S         51.000000  
2      Jyske Bank A/S         51.000000  
3      Jyske Bank A/S         51.000000  
4      Jyske Bank A/S         51.000000  

[5 rows x 21 columns]

In [22]:
data_decline = data[data['Operation Status'] == 'decline']

In [121]:
agg_data = data.groupby(['Customer ID', 'Merchant ID']).agg(
    total=('Customer ID', 'size'),
    total_decline=('Operation Status', lambda x: (x == 'decline').sum()),
    unique_IPs=('Customer IP V4 V6', 'nunique'),
    unique_acc_number=('Account Number', 'nunique'),
    unique_countries=('Country By IP Name', 'nunique'),
    unique_merchants=('Merchant ID', 'nunique'),
    unique_cardholders=('Card Holder', 'nunique'),
    unique_emails=('Customer Email', 'nunique'),
    unique_projects=('Project ID', 'nunique'),
    total_country_missmatch=('Country By IP Name', lambda x: (x != data.loc[x.index, 'Issuer Country Name']).sum())
).reset_index()
agg_data = agg_data.sort_values(by='unique_IPs', ascending=False)


agg_data.head()


Customer ID  Merchant ID  total  total_decline  unique_IPs  \
2055      348231           46     71             16          38   
323       124221         1581     66             59          17   
1903      317704           46     20              4          16   
1966      337528           46     24             21          16   
2006      344501           46     93             34          16   

      unique_acc_number  unique_countries  unique_merchants  \
2055                  1                 2                 1   
323                   1                 2                 1   
1903                  1                 2                 1   
1966                  4                 1                 1   
2006                  2                 2                 1   

      unique_cardholders  unique_emails  unique_projects  \
2055                   1              1                1   
323                    1              1                1   
1903                   1              1                1   
1966                   2              1                1   
2006                   1              1                1   

      total_country_missmatch  
2055                       62  
323                        26  
1903                       18  
1966                        3  
2006                        4

In [126]:
card_usage = data.groupby('Account Number')['Customer ID'].nunique().reset_index()
card_usage.rename(columns={'Customer ID': 'Customer Count'}, inplace=True)

# Идентификация карт, используемых несколькими клиентами
fraudulent_cards = card_usage[card_usage['Customer Count'] > 1]

fraudulent_cards


Account Number  Customer Count
1        '+79192126910               2
407   409858******9159               2
409   409858******9323               2
490   412720******3539               2
495   412720******4466               2
...                ...             ...
4571  558306******0424               2
4599  676867******3691               2
4665       77056207258               2
4853       77718263731               2
5005     Z688880857551               3

[192 rows x 2 columns]

### Все фродовые айпи адреса

In [241]:
data_fraud = [data_fraud1, data_fraud2, data_fraud3]
fraud = pd.concat(data_fraud)['Customer ID']
fraud

35                        8ba8aa0ac77133680329648aaad0109c
59                        8ba8aa0ac77133680329648aaad0109c
95       523927104ac252a76d599ac2c4a5eeb3d0855d6a7a8664...
188      f8b710cfb991a11a883c0f376fe610c1d4d38750ad8154...
204                       8ba8aa0ac77133680329648aaad0109c
                               ...                        
13535    21af7e457c4a882da5fc577226445ac9c6dca700f0e978...
13552    80adaa3dc01dc25c3bcaaa3543a4054ff2821e8d334eba...
13563    5324ac6b2c99504b3011b61e4bfc9444224d810140b2c9...
13564    a8883369c822ec42b9ed804d331c358eb4ba5adb12d198...
13572    c94bbbccb17e1da80d27e9d2897146a8d3b0927d4dee4f...
Name: Customer ID, Length: 1030, dtype: object